File: editor/src/de/project/ice/editor/editors/HolderEditor.java
package de.project.ice.editor.editors;

import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener;
import com.kotcrab.vis.ui.widget.VisTextButton;
import com.kotcrab.vis.ui.widget.VisTextField;
import de.project.ice.utils.Assets;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class HolderEditor<T> extends ValueEditor<Assets.Holder<T>>
{
    private VisTextField valueField;

    @NotNull
    protected String getHolderName()
    {
        if (value != null && value.name != null)
        {
            return value.name;
        }
        return "invalid";
    }

    @Nullable
    protected T getHolderData()
    {
        if (value != null)
        {
            return value.data;
        }
        return null;
    }

    protected void setHolderName(String name)
    {
        if (value != null)
        {
            value.name = name;
        }
    }

    protected void setHolderData(T data)
    {
        if (value != null)
        {
            value.data = data;
        }
    }

    @Override
    protected void createUi()
    {
        valueField = new VisTextField(getHolderName());
        valueField.setDisabled(true);
        add(valueField);
        VisTextButton editButton = new VisTextButton("...", new ChangeListener()
        {
            @Override
            public void changed(ChangeEvent event, Actor actor)
            {
                onEdit();
            }
        });
        add(editButton);
    }

    protected void onEdit()
    {

    }

    @Override
    public void act(float delta)
    {
        super.act(delta);
        valueField.setText(getHolderName());
    }

}

File: editor/src/de/project/ice/editor/editors/HolderEditor.kt
package de.project.ice.editor.editors

import de.project.ice.utils.Assets

abstract class HolderEditor<T> : ValueEditor<Assets.Holder<T>>() {
    private var valueField: VisTextField? = null

    protected val holderName: String
        get() {
            val value = this.value
            if (value != null) {
                return value.name
            }
            return "invalid"
        }

    protected var holderData: Assets.Holder<T>?
        get() {
            if (value != null) {
                return value
            }
            return null
        }
        set(data) {
            if (value != null && data != null) {
                val dataField = Assets.Holder::class.java.getDeclaredField("data")
                dataField.isAccessible = true
                dataField.set(value, data.data)

                val nameField = Assets.Holder::class.java.getDeclaredField("name")
                nameField.isAccessible = true
                nameField.set(value, data.name)
            }
        }

    override fun createUi() {
        valueField = VisTextField(holderName)
        valueField!!.isDisabled = true
        add<VisTextField>(valueField)
        val editButton = VisTextButton("...", object : ChangeListener() {
            override fun changed(event: ChangeListener.ChangeEvent, actor: Actor) {
                onEdit()
            }
        })
        add(editButton)
    }

    protected open fun onEdit() {

    }

    override fun act(delta: Float) {
        super.act(delta)
        valueField!!.text = holderName
    }

}
